Feature/migration to latest dependency versions

jzolnowski opened the pull request:
Migrate dependencies to the latest versions

jzolnowski committed:
feat: Update dependencies to the latest versions, migrate Angular Material to version 15, update configuration in tsconfig and angular.json, improved generation of templates without error related to lack of angular import, replace tslint by eslint and fix all errors

jzolnowski committed:
feat: add pre-commit task to package.json

jzolnowski committed:
feat: add prepare script to install husky

jzolnowski committed:
fix: dependency conflicts

jzolnowski committed:
feat: add toastr AngularJS service, move CDK tree data into Angular service, add BreakpointObserver service and display the currect resolution with Toastr messages

jzolnowski committed:
fix: remove templates folder and import templates directly in app-angularjs.module, add tempaltes to gitignore, fix lint issues. add eslintignore

gkalpak commented on src/app/angular/app.component.ts:
```diff
@@ -1,5 +1,5 @@
+import { BreakpointObserver, BreakpointState, Breakpoints } from '@angular/cdk/layout';
```
:tada:

gkalpak reviewed (dismissed):
~~LGTM~~ :shipit:

_**EDIT:** Sorry, I realized I had only reviewed one commit_ 🤦‍♂️

gkalpak commented on .gitignore:
```diff
@@ -33,6 +33,7 @@ speed-measure-plugin.json
 .history/*
 
 # misc
+/.angular/cache
```
Should we ignore the whole `.angular/` directory instead?

jzolnowski replied:
Fixed

gkalpak commented on src/app/angular/cdk-tree/cdk-tree.component.ts:
```diff
@@ -1,72 +1,8 @@
+import { ArrayDataSource } from '@angular/cdk/collections';
 import { Component } from '@angular/core';
 import { FlatTreeControl } from '@angular/cdk/tree';
```
You missed one 😇

```suggestion
import { FlatTreeControl } from '@angular/cdk/tree';
import { Component } from '@angular/core';
```

jzolnowski replied:
Fixed

gkalpak commented on src/app/angular/cdk-tree/cdk-tree.component.ts:
```diff
@@ -75,21 +11,24 @@ interface ExampleFlatNode {
 
     for (let i = nodeIndex - 1; i >= 0; i--) {
-      if (TREE_DATA[i].level === node.level - 1) {
-        return TREE_DATA[i];
+      if (this.treeData[i].level === node.level - 1) {
+        return this.treeData[i];
       }
     }
```
This can be simplified with the use of `for...of`:

```ts
    for (const otherNode of this.treeData) {
      if (otherNode.level === node.level - 1) {
        return otherNode;
      }
    }
```

jzolnowski replied:
This, unfortunately, will not work the same way. The original loop iterates from the given node upwards and checks if there is a parent, if so, it renders it. Your loop looks for a parent and renders it, but with the leaf omitted. 
Unfortunately, CDK Tree is not working properly, 
<img width="240" alt="image" src="https://user-images.githubusercontent.com/5073455/223482592-75243bcb-6e94-49a2-871e-dcbcbe3da2ad.png">

gkalpak replied:
Sorry, I don't know what I was thinking :grin:
I obviously missed that we are not iterating through the array from the beginning 🤦‍♂️

gkalpak commented on tsconfig.json:
```diff
@@ -15,14 +15,19 @@
     "strictTemplates": true
-  }
+  },
+  "references": [
+    {
+      "path": "./src/tsconfig.app.json"
+    }
+  ]
```
OOC, what is the benefit of using a project reference?

jzolnowski replied:
It allows you to structure your TypeScript programs into smaller pieces. By doing this, you can greatly improve build times, enforce logical separation between components, and organize your code in new and better ways. [Source](https://www.typescriptlang.org/docs/handbook/project-references.html)

gkalpak replied:
Yes, but in this particular case I don't see any difference.
I also remember that the CLI switched to references at some point, but then removed them for some reason (but I don't recall what problem they were causing).

I don't think it makes any difference here (positive or negative).
I am fine either way, but I was trying to understand if there is a specific benefit here that I was missing.

jzolnowski replied:
The benefit of this can be that as a developer you can see all `tsconfig` configuration references for a given application from a single file. After all, you can have many `tsconfig` configurations

gkalpak commented on src/styles.scss:
```diff
@@ -1,22 +1,15 @@
-@include mat.expansion-theme(theme.$app-theme);
-@include mat.progress-spinner-theme(theme.$app-theme);
-@include mat.tabs-theme(theme.$app-theme);
+@include mat.all-component-themes(theme.$app-theme);
```
Does this increase the CSS bundle size? (If so, how much?)

jzolnowski replied:
Yes, calling this mixin increases the size of the bundle. Here are the details:
Within:
<img width="315" alt="image" src="https://user-images.githubusercontent.com/5073455/223484425-e5a173a3-2c5e-45d1-aba1-becba0a1d923.png">

Without:
<img width="310" alt="image" src="https://user-images.githubusercontent.com/5073455/223484468-8f82de6a-0581-48d6-b023-80a3dea2a193.png">

*127kB difference*

We can import thems only for components that are used in the application. What do you think about this @gkalpak?

gkalpak replied:
The increase is not negligible (~25%). I am not as concerned in this demo app as I would be in a real-word app.
Does Angular Material have a guideline on the recommended way to handle this?

jzolnowski replied:
Both solutions are described. When the application does not use too many components, it is written that `mat.all-component-themes` will produce unnecessary CSS.

> As an alternative to listing every component that your application uses, Angular Material offers Sass mixins that include styles for all components in the library: all-component-colors, all-component-typographies, all-component-densities, and all-component-themes. These mixins behave the same as individual component mixins, except they emit styles for core-theme and all 35+ components in Angular Material. Unless your application uses every single component, this will produce unnecessary CSS.

I'm fine with both solutions in the demo app. I would have the same concerns as you if that would be a real-world app 😄

jzolnowski replied:
As we discussed at the meeting, we leave the version `all-component-themes`

gkalpak commented on src/app/angularjs/version-stamp.component.ts:
```diff
@@ -31,20 +30,20 @@ export const versionStampComponent = {
-      <span class="version"><img src="/assets/angular-cdk.svg" alt="Angular CDK icon" height="24px"> Angular CDK v{{ $ctrl.versions.cdk }}</span>
-      <span class="version"><img src="/assets/angular-material.svg" alt="Angular Material icon" height="24px"> Angular Material v{{ $ctrl.versions.mat }}</span>
+      <span class="version"><img src="/assets/angularjs.svg" alt="" height="24px"> AngularJS v{{ $ctrl.versions.angularjs.full }} ({{ $ctrl.versions.angularjs.codeName }})</span>
+      <span class="version"><img src="/assets/angularjs-material.png" alt="" height="24px"> AngularJS Material v{{ $ctrl.versions.md }}</span>
+      <span class="version"><img src="/assets/angular.svg" alt="" height="24px"> Angular v{{ $ctrl.versions.angular }}</span>
+      <span class="version"><img src="/assets/angular-cdk.svg" alt="" height="24px"> Angular CDK v{{ $ctrl.versions.cdk }}</span>
+      <span class="version"><img src="/assets/angular-material.svg" alt="" height="24px"> Angular Material v{{ $ctrl.versions.mat }}</span>
```
Let's also specify the height in CSS.
(BTW, the height property should be set to an [integer value, without a unit][1] :grin:)

[1]: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#:~:text=The%20intrinsic%20height%20of%20the%20image%2C%20in%20pixels.%20Must%20be%20an%20integer%20without%20a%20unit.

gkalpak commented on src/app/angularjs/tabs.component.ts:
```diff
@@ -18,11 +18,15 @@ export const tabsComponent = {
 
 @Directive({ selector: tabsComponent.selector })
-export class TabsComponentFacade extends UpgradeComponent {
+export class TabsComponentDirective extends UpgradeComponent {
```
We have typically called these facades in `ngUpgrade` terminology, because they are just facades for the AngularJS component (they don't have any other logic or purpose).

I don't feel strongly about it, but I do like "Facade", which distinguishes these from regular, general-purpose directives.

jzolnowski replied:
💯 But then you will get lint errors, as the `angular-eslint/recommended` plugin expects "Directive" word as a suffix. We can override the rule. Let us know if we should

<img width="1050" alt="image" src="https://user-images.githubusercontent.com/5073455/223485964-4e3b842d-f196-4daf-a65a-c570a10b7c9b.png">

gkalpak replied:
I see. I don't feel strongly about it, but what I would personally do in hybrid apps (such as this one) is configure the [directive-class-suffix][1] rule to accept both `Directive` and `Facade`. But just using `Directive` is also fine by me. (There's also the option of using `...FacadeDirective` 🤔).

Leaving it up to you - I am OK either way :smiley:

[1]: https://github.com/angular-eslint/angular-eslint/blob/main/packages/eslint-plugin/docs/rules/directive-class-suffix.md

jzolnowski replied:
Changed eslint rules to accept both: `Directive` and `Facade`.
Also changed class names from `TabsComponentDirerctive` to `TabsComponentFacade` and from `VersionStampComponentDirective` to `VersionStampComponentFacade`

gkalpak commented on src/app/angular/app.component.ts:
```diff
@@ -1,11 +1,34 @@
-export class AppComponent {
+export class AppComponent implements OnInit, OnDestroy {
   static selector = 'appComponent';
+  private readonly destroyed: Subject<void> = new Subject<void>();
```
Nit: I generally prefer omitting an explicit type when it can be inferred from the initializer value (but I know we think differently on this one) :grin:

jzolnowski replied:
My opinion comes from Java, where everything, whether the primitive type or not, must have a specific type 😄  I don't insist on it when it comes to TypeScript.

Changed

gkalpak commented on src/app/angularjs/bootstrap-module.ts:
```diff
@@ -1,9 +1,9 @@
 import * as angular from 'angular';
+import { StaticProvider, enableProdMode } from '@angular/core';
+import { MainAngularModule } from '../angular/main-angular.module';
 import { environment } from '../../environments/environment';
 import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
-import { MainAngularModule } from '../angular/main-angular.module';
+import { setAngularJSGlobal } from '@angular/upgrade/static';
```
🙈

jzolnowski replied:
I understand. I'll try to look into automating the sorting of imports with the lint rules 😄

gkalpak reviewed (changes requested):
A few (mostly minor) comments.
Otherwise lgtm :+1:

jzolnowski committed:
Fix issues after George's review

jzolnowski commented:
@gkalpak PR ready for another round of review

gkalpak commented on src/app/angularjs/version-stamp.component.ts:
```diff
@@ -30,11 +31,11 @@ export const versionStampComponent = {
-      <span class="version"><img src="/assets/angular.svg" alt="" height="24px"> Angular v{{ $ctrl.versions.angular }}</span>
-      <span class="version"><img src="/assets/angular-cdk.svg" alt="" height="24px"> Angular CDK v{{ $ctrl.versions.cdk }}</span>
-      <span class="version"><img src="/assets/angular-material.svg" alt="" height="24px"> Angular Material v{{ $ctrl.versions.mat }}</span>
+      <span class="version"><img src="/assets/angularjs.svg" alt="" height="24"> AngularJS v{{ $ctrl.versions.angularjs.full }} ({{ $ctrl.versions.angularjs.codeName }})</span>
+      <span class="version"><img src="/assets/angularjs-material.png" alt="" height="24"> AngularJS Material v{{ $ctrl.versions.md }}</span>
+      <span class="version"><img src="/assets/angular.svg" alt="" height="24"> Angular v{{ $ctrl.versions.angular }}</span>
+      <span class="version"><img src="/assets/angular-cdk.svg" alt="" height="24"> Angular CDK v{{ $ctrl.versions.cdk }}</span>
+      <span class="version"><img src="/assets/angular-material.svg" alt="" height="24"> Angular Material v{{ $ctrl.versions.mat }}</span>
```
Sorry, I think my [previous comment][1] wasn't worded correctly.
I meant we should remove the `height` attribute from the HTML and only specify the height in CSS.

[1]: https://github.com/xlts-dev/angularjs-angular-hybrid-angular-cli/pull/2#discussion_r1124843335

jzolnowski replied:
Removed

gkalpak reviewed (changes requested):
Getting there :wink:

jzolnowski committed:
fix: change class names sufix from Directive to Facade, add @angular-eslint/directive-class-suffix rsrlint rul to accept both: Facade and Directive suffixes, remove height attrribute from versionStamp icons

jzolnowski commented:
@gkalpak PR ready for the final review

jzolnowski committed:
fix: imports order, add comment about usage of all-component-themes mixin

gkalpak commented on src/styles.scss:
```diff
@@ -1,22 +1,15 @@
-@include mat.core(theme.$custom-typography);
+@include mat.core();
 
 // Include theme styles for core and each component used in your app.
```
This line is no longer accurate.
Let's remove it or rephrase.

jzolnowski replied:
Improved comments

gkalpak reviewed (approved):
One minor comment (about a comment :grin:).

LGTM 🚀

jzolnowski committed:
fix: improve comments about Angular Material theming

jzolnowski commented:
@gkalpak PR ready for the final review

jzolnowski committed:
Import-order rule configuration

jzolnowski committed:
fix: imports order

gkalpak reviewed (approved):
Yay for sorted imports :tada:

jzolnowski committed:
feat(test): playwright configuration and e2e tests (#3)

- fix cdk tree component
- fix lint issues
- add .angular folder to .gitignore
- remove templates from source control
- add templates to gitignore
- add console.log error test
- add tab content visibility test
- replace waitForTimeout with waitForLoadState
- improve playwright configuration
- change assertion in a GitHub link test
- replace click handlers in Page Objects with direct `click()` method invocations
- simplify `shouldRender` method in `cdk-tree` component
- add test packackage.json script

Splaktar commented:
Please rebase and resolve conflicts.

jzolnowski commented:
Branch up to date with master.
Additionally:
- fixed the console warning about changing the target to es2022
- improved the styles in version-stamp.component

Splaktar commented on package.json:
```diff
@@ -20,66 +21,82 @@
+    "typescript": "~4.9.4"
   },
   "husky": {
     "hooks": {
```
The engines below is set to npm v6 only. But you've updated the package-lock.json file using npm v8. We either need to update this PR to use only npm v6, or update the engines field below if we're sure that the updates to dependencies here all work with npm v8. If that's the case, we should also update the `.npmrc` file to at least 16 and verify that works.

jzolnowski replied:
I set the `engine-strict` flag in `.npmrc`. I updated the npm and node versions in the `package.json` engines: `npm v9`, `node>=18`. I tested if the package installation, start and testing of the application works correctly. Everything looks fine 🎉

gkalpak replied:
:tada:

Personally, I wouldn't go as far as enabling `engine-strict`. The rest lgtm :+1:

jzolnowski replied:
Disabled `engine-stric` flag

jzolnowski committed:
chore: set engine-strict to true, set engine npm version to 9 and node to above or equal 18

Commit:
chore: update .nvmrc to use node v18

Commit:
chore: remove engine-strict flag

gkalpak commented:
@jzolnowski, make sure all commits are signed :pray:

Commit:
fix: version-stamp.component styles, remove lint job from pre-commit
package.json script"

jzolnowski commented:
> @jzolnowski, make sure all commits are signed 🙏

All commits signed 🎉

gkalpak commented on src/app/angular/cdk-tree/cdk-tree.component.ts:
```diff
@@ -75,29 +11,32 @@ interface ExampleFlatNode {
-  dataSource = new ArrayDataSource(TREE_DATA);
+  constructor(private cdkTreeService: CdkTreeService) {}
+
+  dataSource = new ArrayDataSource(this.treeData);
```
Nit: Properties (at least non-function ones) should go above the constructor.

jzolnowski replied:
Changed

gkalpak commented on package.json:
```diff
@@ -5,6 +5,7 @@
   "private": true,
   "scripts": {
     "ng": "ng",
+    "pre-commit": "lint-staged",
```
Don't we need a config for `lint-staged`?

jzolnowski replied:
I'm not sure if you mean the configuration in package.json, but we have one [here](https://github.com/xlts-dev/angularjs-angular-hybrid-angular-cli/pull/2/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R106) 😄

gkalpak replied:
Wow! Not sure how I missed that 🤦‍♂️

gkalpak reviewed (changes requested):
A side comment:
This PR does too much. In the future, it would be better to break it up into multiple PRs.
Also, related but independent, self-contained changes in the same PR can be organized in separate commits.

This makes it easier to review and merge PRs.

Commit:
chore: move class field above the constructor

Commit:
fix(lint): Failed to load plugin 'import' declared in '.eslintrc issue

gkalpak commented on package.json:
```diff
@@ -20,66 +21,83 @@
+    "eslint-config-prettier": "8.6.0",
+    "eslint-import-resolver-typescript": "^2.5.0",
+    "eslint-plugin-ban": "^1.6.0",
+    "eslint-plugin-import": "^2.27.5",
```
Is this needed?

jzolnowski replied:
I do not know why, but the previous time I had an error lint:

<img width="906" alt="image" src="https://user-images.githubusercontent.com/5073455/230589438-9a024cf6-ac9b-4674-8087-018e964aa9ea.png">

I see that it has disappeared. I have removed the dependency

Commit:
chore: set package.json engine.node field to v16, generate new package-lock-json, set nvmrc node to v16, remove no needed eslint-plugin-import dependency

jzolnowski commented:
`engines.node` field in `package.json` set to v16

gkalpak reviewed (approved):
LGTM :+1:
